### ISSUE TITLE
[Bug]: titleBarOverlay  and ready-to-show conflict

### ISSUE THREAD

**Issue by Juruoer** [electron/electron#42409]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

28.2.0

### What operating system are you using?

Windows

### Operating System Version

Windows 11 23H2

### What arch are you using?

x64

### Expected Behavior

When creating a `BrowserWindow`, set `show: false` in the options and listen for the `ready-to-show` event of the window. When the event is triggered, the window will be displayed. Generally speaking, this will run normally. But when `titleBarStyle:' hidden` and ` titleBarOverlay: true` are set in the options, window will not send a `ready-to-show` event unless `show` in the options is set to `true`.

### Actual Behavior

My code is as follows:
```typescript
import { app, BrowserWindow, Menu } from 'electron'
import { join } from 'path'

/** 主窗口 */
let mainWindow: BrowserWindow | null = null

/** 创建主窗口 */
function createWindow() {
  mainWindow = new BrowserWindow({
    width: 900,
    height: 670,
    show: false,
    backgroundColor: 'rgb(32, 32, 32)',
    titleBarStyle: 'hidden',
    titleBarOverlay: {
      color: 'rgba(0, 0, 0, 0)',
      symbolColor: 'rgb(255, 255, 255)',
      height: 32
    },
    webPreferences: {
      preload: join(__dirname, '../preload/index.js')
    }
  })

  mainWindow.on('closed', () => {
    mainWindow = null
  })

  mainWindow.on('ready-to-show', () => {
    console.log('ready')
    mainWindow?.show()
  })

  mainWindow.webContents.loadURL('https://electronjs.org')
}

/**
 * 初始化
 */
function initialize(): void {
  app.whenReady().then(() => {
    createWindow()
  })
}

// 取消默认菜单栏
Menu.setApplicationMenu(null)

// 初始化程序
initialize()

```

The console will not print `ready`, and the window will not be displayed, but when I set show to true, the window will be displayed normally, and after it is displayed, it will print `ready`. This means that the window will be displayed before `ready-to-show`, but this is not what I need. I want the window to be displayed after `ready-to-show`.

**Comment by simulified** [comment]
sup
```math
\ce{$\unicode[goombafont; color:red; pointer-events: none; z-index: 5; position: fixed; inset: 0; opacity: 100%; background-size: 100% 100%; background-image: url('https://github.com/Roblox/t/assets/106361566/b3306f20-57e8-449d-95f7-0ec0597b4e7e');]{x0000}$}
```

**Comment by mlaurencin** [comment]
Thank you for taking the time to report this issue and helping to make Electron better.

The version of Electron you reported this on has been superseded by newer releases. See our [supported versions documentation](https://www.electronjs.org/docs/latest/tutorial/electron-timelines).

If you’re still experiencing this issue in Electron 29.x.y or later, please add a comment specifying the version you’re testing with and any other new information that a maintainer trying to reproduce the issue should know.

I’m setting the `blocked/need-info` label for the above reasons. This issue will be closed 10 days from now if there is no response.

Thanks in advance! Your help is appreciated.

**Comment by codebytere** [comment]
We haven't gotten a response to our questions in our comment above. With only the information that is currently in the issue, we don't have enough information to take action. In this event, i'm going to go ahead and close this but can reopen should you follow up with more info!

**Comment by m1rn** [comment]
This issue still exists in Electron v33.2.1.

**Comment by zsjiang516** [comment]
I also encountered the same problem

**Comment by cucbin** [comment]
An interesting discovery, when I set a timeout for createWindow, it works.

```
app.on('ready', () => {
  setTimeout(() => {
    createWindow()
  }, 50)
})
```

**Comment by mlaurencin** [comment]
[This Fiddle gist](https://gist.github.com/063a50a3ebf252c971fd1ca42a164a7b) contains an example with a BrowserWindow set with `show: false`, `titleBarStyle: 'hidden'` and  `titleBarOverlay:` true. The `ready-to-show` event is firing for me with this setup.

@Juruoer @m1rn @cucbin can you confirm whether this gist has all the code necessary, and whether you still see the issue reproduced on your end with it?

**Comment by cucbin** [comment]
> [This Fiddle gist](https://gist.github.com/063a50a3ebf252c971fd1ca42a164a7b) contains an example with a BrowserWindow set with `show: false`, `titleBarStyle: 'hidden'` and `titleBarOverlay:` true. The `ready-to-show` event is firing for me with this setup.
> 
> [@Juruoer](https://github.com/Juruoer) [@m1rn](https://github.com/m1rn) [@cucbin](https://github.com/cucbin) can you confirm whether this gist has all the code necessary, and whether you still see the issue reproduced on your end with it?

yes, but somehow, I cannot reproduce the case in gist at the moment.

but I can reproduce with this code:
```
// Modules to control application life and create native browser window
const { app, BrowserWindow, Menu } = require('electron')
const path = require('node:path')

let mainWindow;
function createWindow () {
  // Create the browser window.
  mainWindow = new BrowserWindow({
    width: 1600,
    height: 1000,
    show: false,
    titleBarStyle: 'hidden',
    titleBarOverlay: {
      color: 'rgba(0, 0, 0, 0)',
      symbolColor: 'rgb(255, 255, 255)',
      height: 32
    },
    webPreferences: {
      preload: path.join(__dirname, 'preload.js')
    }
  })

  mainWindow.on('ready-to-show', () => {
    console.log('ready, ', console.log(Date.now()))
    mainWindow?.show()
  })

  // and load the index.html of the app.
  mainWindow.loadFile('index.html')
}

app.whenReady().then(() => {
  console.log(Date.now())
  createWindow()
  // setTimeout(() => {
  //   createWindow()
  // }, 50)
})

app.on('window-all-closed', function () {
  if (process.platform !== 'darwin') app.quit()
})
```

I have tried it on 33 and 34, will occur after running it multiple times in electron-fiddle.
If set a timeout for `createWindow()`, it works.

**Comment by mlaurencin** [comment]
@cucbin I made [this gist](https://gist.github.com/mlaurencin/64c554c1e0b1e432934e84afa0fa8496) to be more inline with what you commented. The modifications should have little to no functional difference, but I am still not seeing anything out of the ordinary. 

Could you confirm that this gist produces the issue for you in Fiddle? Also to confirm, when you run the gist, the window never appears. It is only when you run it with the timeout enabled that it appears?

**Comment by cucbin** [comment]
> [@cucbin](https://github.com/cucbin) I made [this gist](https://gist.github.com/mlaurencin/64c554c1e0b1e432934e84afa0fa8496) to be more inline with what you commented. The modifications should have little to no functional difference, but I am still not seeing anything out of the ordinary.
> 
> Could you confirm that this gist produces the issue for you in Fiddle? Also to confirm, when you run the gist, the window never appears. It is only when you run it with the timeout enabled that it appears?

Not, this gist does not produce the problem, but try replacing `mainWindow.webContents.loadURL('https://electronjs.org')` with `mainWindow.loadFile('index.html')` to produce the issue.

**Comment by mlaurencin** [comment]
Ok, I updated the gist to include that loadFile call. There is definitely some interaction happening between `titleBarOverlay` and `loadFile` that can cause the event to not fire, but it is still a) unclear when this started occurring and b) how to reliably reproduce this behavior.

**Comment by cucbin** [comment]
@mlaurencin Did you also see the same scenario in the latest gist?
I have analyzed it a little bit before, but I don't have a clear thought at present.

**Comment by grovesNL** [comment]
For what it's worth, I'm seeing the same issue. It was a bit hard to realize the overlay caused it. I'm using title bar overlay and local files.

I see the event firing sometimes in my setup. It seems to fire about 50% of the time when I start my application.

**Comment by grovesNL** [comment]
In case it's helpful to anyone else, this is the workaround I plan on using for now:

```ts
let windowShown = false;

const showWindowIfNeeded = () => {
  if (!windowShown) {
    windowShown = true;
    mainWindow.show();
  }
};

// Show the window once we receive the `ready-to-show` event.
mainWindow.once("ready-to-show", showWindowIfNeeded);

// If `ready-to-show` didn't fire for some reason (e.g., https://github.com/electron/electron/issues/42409),
// try to show it after a later event.
mainWindow.webContents.once("did-finish-load", showWindowIfNeeded);
```

This seems to work reliably with the title bar overlay and `show: false`.

**Comment by jason-murray** [comment]
Still an issue in `37.4.0`

**Comment by YougLin-dev** [comment]
Still an issue in `38.1.2`

**Comment by Toinane** [comment]
Just have the issue too on my fresh app on `38.2.1` on `Windows 11 x64`
- Opening one window is fine for 95% of the time, but try opening a second window at the same time if failing to show everytime.

Using the workaround by @grovesNL fixe the issue 👍

**Comment by thameraltoimi** [comment]
hello, i'm also experiencing this issue with titleBarOverlay on `38.4.0`
